File: examples/monitor.c
#include <gautogui/gautogui.h>

static void
on_mouse_moved(GAutoguiController *controller,
               gint x,
               gint y,
               gpointer user_data)
{
  (void)controller;
  (void)user_data;

  g_print("mouse moved: x=%d y=%d\n", x, y);
}

static void
on_mouse_button_event(GAutoguiController *controller,
                      guint button,
                      gboolean pressed,
                      gint x,
                      gint y,
                      gpointer user_data)
{
  (void)controller;
  (void)user_data;

  g_print("mouse button: button=%u %s x=%d y=%d\n",
          button,
          pressed ? "down" : "up",
          x,
          y);
}

static void
on_mouse_scrolled(GAutoguiController *controller,
                  gint dx,
                  gint dy,
                  gint x,
                  gint y,
                  gpointer user_data)
{
  (void)controller;
  (void)user_data;

  g_print("mouse scrolled: dx=%d dy=%d x=%d y=%d\n", dx, dy, x, y);
}

static void
on_key_event(GAutoguiController *controller,
             guint native_keycode,
             guint key,
             gboolean pressed,
             gpointer user_data)
{
  (void)controller;
  (void)user_data;

  g_print("key: native=%u key=%u %s\n",
          native_keycode,
          key,
          pressed ? "down" : "up");
}

int
main(void)
{
  g_autoptr(GAutoguiController) controller = NULL;
  g_autoptr(GMainLoop) loop = NULL;
  g_autoptr(GError) error = NULL;

  controller = gautogui_controller_new();
  g_signal_connect(controller, "mouse-moved", G_CALLBACK(on_mouse_moved), NULL);
  g_signal_connect(controller, "mouse-button-event", G_CALLBACK(on_mouse_button_event), NULL);
  g_signal_connect(controller, "mouse-scrolled", G_CALLBACK(on_mouse_scrolled), NULL);
  g_signal_connect(controller, "key-event", G_CALLBACK(on_key_event), NULL);

  if (!gautogui_controller_start(controller, &error)) {
    g_printerr("Unable to start gautogui monitor: %s\n", error->message);
    return 1;
  }

  g_print("Monitoring system-wide input. Press Ctrl+C to quit.\n");

  loop = g_main_loop_new(NULL, FALSE);
  g_main_loop_run(loop);

  return 0;
}

File: examples/monitor.py
#!/usr/bin/env python3

import sys

import gi

gi.require_version("Gautogui", "1.0")

from gi.repository import GLib, Gautogui


def on_mouse_moved(controller, x, y):
    print(f"mouse moved: x={x} y={y}")


def on_mouse_button_event(controller, button, pressed, x, y):
    state = "down" if pressed else "up"
    print(f"mouse button: button={button} {state} x={x} y={y}")


def on_mouse_scrolled(controller, dx, dy, x, y):
    print(f"mouse scrolled: dx={dx} dy={dy} x={x} y={y}")


def on_key_event(controller, native_keycode, key, pressed):
    state = "down" if pressed else "up"
    print(f"key: native={native_keycode} key={key} {state}")


def main():
    controller = Gautogui.Controller.new()
    controller.connect("mouse-moved", on_mouse_moved)
    controller.connect("mouse-button-event", on_mouse_button_event)
    controller.connect("mouse-scrolled", on_mouse_scrolled)
    controller.connect("key-event", on_key_event)

    try:
        controller.start()
    except GLib.Error as exc:
        print(f"Unable to start gautogui monitor: {exc.message}", file=sys.stderr)
        return 1

    print("Monitoring system-wide input. Press Ctrl+C to quit.", flush=True)

    loop = GLib.MainLoop()
    try:
        loop.run()
    except KeyboardInterrupt:
        controller.stop()

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
